File: rust/src/solutions/day06.rs
// What did I learn?
// contains, some slice/sub-slce manipulation. Used directly u8 instead of chars for simplicity.

use log::info;
use crate::Solver;

pub(crate) struct Solution {
    start_of_packet: Option<usize>,
    start_of_message: Option<usize>,
}

impl Solution {
    pub(crate) fn new() -> Solution {
        Solution {
            start_of_packet: None,
            start_of_message: None,
        }
    }
}

fn all_different(buf: &[u8], start: usize, length: usize) -> bool {
    let end = start + length;
    for i in start..end - 1 {
        let x = &buf[i];
        let found = buf[i + 1..end].contains(x);
        if found {
            return false;
        }
    }
    return true;
}

impl Solver for Solution {
    fn parse(&mut self, line: &str) {
        let mut i: usize = 0;
        while i + 14 <= line.len() {
            let bytes = line.as_bytes();
            if self.start_of_packet.is_none() && all_different(bytes, i, 4) {
                self.start_of_packet = Some(i + 4);
                info!("[1] Found start of packet at {}", i + 4);
            }
            if self.start_of_message.is_none() && all_different(bytes, i, 14) {
                self.start_of_message = Some(i + 14);
                info!("[2] Found start of message at {}", i + 14);
            }
            if self.start_of_packet.is_some() && self.start_of_message.is_some() {
                return;
            }
            i += 1;
        }
    }

    fn solve(&mut self) -> Option<(String, String)> {
        Some((self.start_of_packet.unwrap().to_string(), self.start_of_message.unwrap().to_string()))
    }
}

File: python/src/adventofcode/year2022/day06.py
import logging

from adventofcode import Solver


# https://adventofcode.com/2022/day/6
# This was actually "simple" once one realises that flinging a slice into a set would create a set of the expected
# length only if all letters were different.
# one could actually use a single dictionary containing frequency of letters and add/remove as the sliding window moves
# but the code would end up being more involved.
# the `solve()` is an empty method here as the input is a single line

log = logging.getLogger("day.06")

def all_different(string):
    return len(set(string)) == len(string)


class Solution(Solver):
    def __init__(self):
        self.start_of_packet = None
        self.start_of_message = None

    def parse(self, line: str):
        start_of_packet = False
        start_of_message = False
        line = line.strip()
        i = 0
        while i + 14 <= len(line):
            # start_of_packet : 4 chars
            if not self.start_of_packet and all_different(line[i:i+4]):
                self.start_of_packet = i + 4
                log.info(f"Found start-of-packet at {self.start_of_packet}")
            if not self.start_of_message and all_different(line[i:i+14]):
                self.start_of_message = i + 14
                log.info(f"Found start-of-message at {self.start_of_message}")
            if start_of_packet and start_of_message:
                return
            i += 1

    def solve(self):
        return str(self.start_of_packet), str(self.start_of_message)
